### PR TITLE
[SYCL] Allow batching for wait with a barrier commands

### PR DESCRIPTION
**Pull request by smaslov-intel**
Signed-off-by: Sergey V Maslov <sergey.v.maslov@intel.com>

### CODE REVIEW

File: sycl/plugins/level_zero/pi_level_zero.cpp [hunk 1]
```diff
@@ -5379,6 +5379,22 @@ pi_result piEnqueueEventsWaitWithBarrier(pi_queue Queue,
   PI_ASSERT(Queue, PI_INVALID_QUEUE);
   PI_ASSERT(Event, PI_INVALID_EVENT);
 
+  // Submit dependent open command lists for execution, if any
```
Comment (againull) [inline sycl/plugins/level_zero/pi_level_zero.cpp:5382]: Could you please clarify in the comment why we need to execute all command lists.
It doesn't look obvious why it is required.

Also this maybe not closely related to this PR but how we deal with the situation when events in the waitlist are associated with device(s) which are different from device where command list is created. It looks like it is illegal per L0 spec:
https://spec.oneapi.io/level-zero/latest/core/api.html?highlight=zecommandlistappendbarrier#zecommandlistappendbarrier

> The application must ensure the events are accessible by the device on which the command list was created.

In opencl requirement is less strict and it is enough for queue and events to be in the same context:
https://www.khronos.org/registry/OpenCL/sdk/2.2/docs/man/html/clEnqueueBarrierWithWaitList.html

> The context associated with events in event_wait_list and command_queue must be the same.

It doesn't look like we have a guarantee that all events in the list are associated with the same device or its subdevices.

Comment (smaslov-intel) [reply sycl/plugins/level_zero/pi_level_zero.cpp:5382]: > why we need to execute all command lists.

This is to ensure forward progress, and no deadlock.
If we don't submit commands being waited, then the wait may be infinite.

> deal with the situation when events in the waitlist are associated with device(s) which are different from device where command list is created

That's the same as for all other commands, e.g. see zeCommandListAppendLaunchKernel:
https://spec.oneapi.io/level-zero/latest/core/api.html?highlight=zecommandlistappend#_CPPv431zeCommandListAppendLaunchKernel24ze_command_list_handle_t18ze_kernel_handle_tPK16ze_group_count_t17ze_event_handle_t8uint32_tP17ze_event_handle_t

If this is an issue then we'd address it separately, and I don't know how at this moment.

Comment (againull) [reply sycl/plugins/level_zero/pi_level_zero.cpp:5382]: > This is to ensure forward progress, and no deadlock.
> If we don't submit commands being waited, then the wait may be infinite.

Indeed, got it, thank you.

> That's the same as for all other commands, e.g. see zeCommandListAppendLaunchKernel:

You are right, I didn't notice this until now.

**Review by againull — APPROVED** [review]

**Pull request merged**